AX: Collapse AccessibilityLabel into AccessibilityRenderObject

hoffmanjoshua opened the pull request:
#### 6deff2be0ce143a057e4319a1a99976a1f6851af
<pre>
AX: Collapse AccessibilityLabel into AccessibilityRenderObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=297634">https://bugs.webkit.org/show_bug.cgi?id=297634</a>
<a href="https://rdar.apple.com/158730216">rdar://158730216</a>

Reviewed by Tyler Wilcock.

Collapses all of the logic from AccessibilityLabel into AccessibilityRender- and
AccessibilityNodeObject as applicable.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::shouldCreateAccessibilityLabel): Deleted.
* Source/WebCore/accessibility/AccessibilityLabel.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityLabel.h: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::clearChildren):
(WebCore::AccessibilityNodeObject::textAsLabelFor const):
(WebCore::AccessibilityNodeObject::textForLabelElements const):
(WebCore::childrenContainOnlyStaticText):
(WebCore::AccessibilityNodeObject::isLabelContainingOnlyStaticText const):
(WebCore::AccessibilityNodeObject::isNativeLabel const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityTableColumnInstance const):
(WebCore::AccessibilityObject::isNativeLabel const):
(WebCore::AccessibilityObject::isLabelByRelation const):
(WebCore::AccessibilityObject::isLabel const):
(WebCore::AccessibilityObject::isAccessibilityLabelInstance const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::stringValue const):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: <a href="https://commits.webkit.org/299008@main">https://commits.webkit.org/299008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e0f3d911ccb496910941b11498735497c2116b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117382 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123481 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69370 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b1dab157-6d6f-474e-9733-74227cab95ff) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119260 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89072 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43741 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/35c6b14d-9f5e-454e-9598-d7e7780fb13f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30041 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105249 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69580 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1f27a745-d6ce-4e5d-949e-e782d90edad5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29104 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67154 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99444 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23547 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126602 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44279 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33272 "Found 1 new test failure: ipc/send-gradient.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97748 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44636 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101485 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97542 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24847 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42902 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20839 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40629 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44152 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49811 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43608 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46953 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45304 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7d4a8aa8c2b07542d8149e83be77a1a82b4bb7af. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49638

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -155,6 +158,11 @@ class AccessibilityNodeObject : public AccessibilityObject {
     // Only used by AccessibilitySVGObject, but placed here to use space that would otherwise be taken by padding.
     bool m_isSVGRoot { false };
+
+    // Only used by labels, but placed here (rather than rare data) to use space that would otherwise be taken by padding.
```
"Only used by labels" doesn't feel accurate to me, since there is no AccessibilityLabel anymore. Maybe "Only used by objects representing labels" or something like that? Also I don't think the rare data callout is necessary (it doesn't even exist in `main` yet)

hoffmanjoshua replied:
Updated this comment in my latest commit

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -673,6 +672,10 @@ void AccessibilityNodeObject::clearChildren()
 {
     AccessibilityObject::clearChildren();
     m_childrenDirty = false;
+    if (isLabel()) {
+        m_containsOnlyStaticText = false;
+        m_containsOnlyStaticTextDirty = false;
+    }
```
Prior to this commit, there were two types of "labels", and only one needed to manage these two stateful bools.

  1. Whether something is `labelForObjects().size()`
  2. Whether something was an AccessibilityList

The `labelForObjects().size()` does not and should not be writing or respecting these bools, right? But with this commit, we've conflated these two types into one `isLabel()` concept. This has downsides, like in this case, where we add a virtual function call to a very hot function, `clearChildren`. It also seems like a pretty big change in behavior to have anything `labelForObjects().size()` now start reading and writing these bools, which impacts things like computed roles.

hoffmanjoshua replied:
Fixed: we have two different methods now. `isLabel` and `isLabelForObject`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3058,6 +3061,45 @@ bool AccessibilityNodeObject::isDescriptionList() const
+    return true;
+}
+
+bool AccessibilityNodeObject::labelContainsOnlyStaticText() const
```
Something like `isLabelContaininingOnlyStaticText` might be more accurate? It also seems like we'll be duplicating the `isLabel()` check in the only place that calls this new method? Probably not the biggest deal though

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/189a4e55216e5d15183e850582844acc7265725b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49638

twilco commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -134,7 +133,8 @@ class AccessibilityObject : public AXCoreObject {
     virtual bool isSliderThumb() const { return false; }
-    bool isLabel() const { return isAccessibilityLabelInstance() || labelForObjects().size(); }
+    virtual bool isLabel() const { return false; }
+    bool isLabelForObject() const;
```
Any reason this isn't just inlined?

`bool isLabelForObject() const { return labelForObjects().size(); }`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -189,6 +192,11 @@ class AccessibilityNodeObject : public AccessibilityObject {
     // Only used by AccessibilitySVGObject, but placed here to use space that would otherwise be taken by padding.
     bool m_isSVGRoot { false };
+
+    // Only used by objects representing labels to use space that would otherwise be taken by padding.
```
Hmm, this sentence may not read clearly? "Only used by objects representing labels to use space"

If we take my other suggestion, we could say something like:

`// Only used by isNativeLabel() objects. Placed here to use space that would otherwise be taken by padding.`

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6b488a73679e633fdcf91e640c665e74bf2ff223. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49638

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1368,7 +1357,7 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
         }
 
-        if (parent->isLabel()) {
+        if (parent->isNativeLabel()) {
```
Current `main` behavior runs this code when:

`bool isLabel() const { return isAccessibilityLabelInstance() || labelForObjects().size(); }`

Don't we need to run this branch when `isLabelByRelation()` as well to maintain the existing behavior? I'd assume some tests would start failing when we only do this for `isNativeLabel()` 🤔

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7e0f3d911ccb496910941b11498735497c2116b7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49638

hoffmanjoshua committed:
AX: Collapse AccessibilityLabel into AccessibilityRenderObject
https://bugs.webkit.org/show_bug.cgi?id=297634
rdar://158730216

Reviewed by Tyler Wilcock.

Collapses all of the logic from AccessibilityLabel into AccessibilityRender- and
AccessibilityNodeObject as applicable.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::shouldCreateAccessibilityLabel): Deleted.
* Source/WebCore/accessibility/AccessibilityLabel.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityLabel.h: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::clearChildren):
(WebCore::AccessibilityNodeObject::textAsLabelFor const):
(WebCore::AccessibilityNodeObject::textForLabelElements const):
(WebCore::childrenContainOnlyStaticText):
(WebCore::AccessibilityNodeObject::isLabelContainingOnlyStaticText const):
(WebCore::AccessibilityNodeObject::isNativeLabel const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityTableColumnInstance const):
(WebCore::AccessibilityObject::isNativeLabel const):
(WebCore::AccessibilityObject::isLabelByRelation const):
(WebCore::AccessibilityObject::isLabel const):
(WebCore::AccessibilityObject::isAccessibilityLabelInstance const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::stringValue const):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: https://commits.webkit.org/299008@main
